### ISSUE TITLE
[proposal] Extend WeavingHook to cover resource loading as well

### ISSUE THREAD

**Issue by laeubi** [osgi/osgi#278]
OSGi Framework Weaving Hook Service is described as:

> Bundles registering this service will be called during framework class loading operations. Weaving hook services are called when a class is being loaded by the framework and have an opportunity to transform the class file bytes that represents the class being loaded.

while this is a powerful tool, there is one aspect missing:

A classlaoder allows beside finding classes also finding of [resources](https://docs.oracle.com/javase/7/docs/api/java/lang/ClassLoader.html#findResource(java.lang.String)). Many frameworks use this approach to discover configurations, meat-data files, images and so on.

My proposal here is extend the WeavingHook (e.g. via a default interface method or additional interface to implement) with a method:

```
public default void weave(WovenResource wovenResource) {
  //default implementation do nothing
}
```

A WovenResource resource is created and handed over to the hook whenever a resource is requested by the bundle-classloader. Similar to the WovenClass it has to carry some context information and provide the following actions:

- read the resource (similar to WovenClass#getBytes())
- replace the resource (similar to WovenClass#setBytes(...))
- hide the resource (similar to WeavingHook#weave throwing a WeavingException)

as a resource might only be discovered (in the form of a URL) but only read later, the operation of replacing a resource might offer a way for a lazy content generation (e.g. via `Supplier<InputStream>`) and read of the resource might be delayed upon an actual call to getBytes().

Similar to how class weaving currently works WeavingHooks that are called later will see already transformed data of previous WeavingHooks.

With this in place, advanced transformations are possible, e.g. replacing placeholders, choosing configurations based on the running os, or even placing empty dummy resources in a bundle that gets gets generated at runtime without the need to modify the codepath in the application or enhance existing libraries to support special extension points where custom code could hook into.

**Comment by tjwatson** [comment]
Once a class is defined it effectively becomes immutable and a WeavingHook will not see a WovenClass object for the same class loader/class name pair (assuming no class unloading is going on).  Would the same hold true for resource weaving?  Resources can be read multiple times and in the case of exported packages they may be read by multiple bundles.  Once a resource has been woven would the resulting content need to be cached by the framework so that others reading the resource would be guaranteed to see the same result or are we going to allow a WeavingHook to change the content each call to getResource?

What about resources in non-package directories (e.g. /META-INF folder).  I think it would be dangerous to consider resources such as these to be woven.  One may think they can weave the manifest to augment the bundle meta-data, but this would be far too late because by this point we have a resolved bundle with a class loader all based on the real manifest.  Also, we should not consider weaving other stuff in the META-INF folder like signer data.

It could get confusing to differentiate woven resources obtained by the class loader and their corresponding resources that may be obtained by Bundle.getEntry without going through the class loader.  I assume Bundle.getEntry resources would never be woven.

**Comment by bjhargrave** [comment]
Yes, I am not sure this is a good idea. Weaving is primarily about modifying code to weave in some cross-cutting behavior. Once a class is loaded by a class loader the byte array is no longer interesting as it is converted to a Class object which is immutable. Resources are fundamentally different. They are not transformed by a class loader into some immutable form (e.g. Class object). So "weaving" resources would either require them to be mutated on each request (which is problematic in performance and consistency terms) or require the framework to cache the resulting woven resource to return in the future (this also has performance issues as it absorbs memory almost like a memory leak as it is possible that no one will re-request the resource).

So I don't think this is a good idea.

**Comment by laeubi** [comment]
> Once a class is defined it effectively becomes immutable and a WeavingHook will not see a WovenClass object for the same class loader/class name pair (assuming no class unloading is going on). Would the same hold true for resource weaving?

A resources might be read multiple times but this will always result in the same result, so resources are kind of static also. 

Actually we are talking about `java.lang.ClassLoader.getResources(String)` /  `java.lang.ClassLoader.getResource(String)` and I think those are well defined that there is no guarantee of caching but also do not forbids caching if desired.

So a BundleClassLoader supporting "ResourceWeaving" would work the following way:

1. implements `java.lang.ClassLoader.findResource(String)` as of today and will return an URL to a special protocol that is capable of reading a resource from the bundles classpath
2. If the URL is about to be opened will create a `WeavingURLConnection` with encapsulates a WovenResource initialized with a `Supplier<InputStream>` that can supply the data as it is done today
3. If there are WeavingHooks, pass that WovenResource to the Hooks and they can set the `Supplier<InputStream>` to a different one
4. if later the openStream is called it simply delegates to the supplier

> Weaving is primarily about modifying code to weave in some cross-cutting behavior.

And the same can apply to resources as well, how is this different? Lets say I'd like AspektJ or similar to add automatic logging to every class from the bussinesslayer, why should I not be allowed to replace a variable in all logging.properties files or alike?

> Once a class is loaded by a class loader the byte array is no longer interesting

The same applies to URLConnections beside that they have a much shorter life-cycle and a higher chance of freeing resources again

> Resources are fundamentally different. They are not transformed by a class loader into some immutable form (e.g. Class object)

It clearly depends how they are used, but I disagree that they are fundamentally different (they are just not that limited), if you take for example a function that reads icons from the classpath, these would be read once, decoded transferred into some kind of in-memory representation of an pixel array and then be used unless they are disposed again. The same will apply to config files, I'm not aware of any framework that constantly parses a textual config file over and over again, they will be read once and stored in an internal representation (e.g. Properties object).

> would either require them to be mutated on each request [...] or require the framework to cache the resulting woven resource to return in the future

I don't see any need for caching nor problematic access as there is no much difference to today, a framework might choose to cache resources (to prevent disk I/O) or read it every time it is requested. If transforming adds a penalty clearly depends on the type of transformation and is out of the scope of the framework or this request.

> What about resources in non-package directories (e.g. /META-INF folder).

As described above I would only consider resources that are loaded through classloaders.

> It could get confusing to differentiate woven resources obtained by the class loader and their corresponding resources that may be obtained by Bundle.getEntry without going through the class loader. I assume Bundle.getEntry resources would never be woven.

As of today, all these Bundle.getXXX methods are well described whether or not the BundleClasslaoder is used or not so there actually already is a difference. I would simply add a comment to those not using the classloader that ResourceWeaving would not happen.

> So I don't think this is a good idea.

Let me know your concerns so we can discuss them, from the once mentions until now many mechanism we find in modern OSGi Frameworks should have never been included (e.g. subsystems and alike), this one is really nothing much to harm...

**Comment by tjwatson** [comment]
> I have given detailed remarks for the concerns, simply declaring a valid demand as "not a wise idea" makes feeling the project do not like contributions or is interested in the requirements of its consumers 👎

That certainly is not the intent.

> > Once a class is defined it effectively becomes immutable and a WeavingHook will not see a WovenClass object for the same class loader/class name pair (assuming no class unloading is going on). Would the same hold true for resource weaving?
> 
> A resources might be read multiple times but this will always result in the same result, so resources are kind of static also.

But if weavers are involved then the content read may change over time as weavers are registered/unregistered.  Once a class is defined, that is it until the bundle is refreshed and we throw away the class loader.  If a weaver comes along after a class is defined it cannot change the current class.  Yet if we allow weavers to come along to transform resources I fear the transformation could get out of sync with what the weaver would do for the already defined classes.  But there is no way for the framework to know what resources go with what classes.

> 
> Actually we are talking about `java.lang.ClassLoader.getResources(String)` / `java.lang.ClassLoader.getResource(String)` and I think those are well defined that there is no guarantee of caching but also do not forbids caching if desired.
> 
> So a BundleClassLoader supporting "ResourceWeaving" would work the following way:
> 
> 1. implements `java.lang.ClassLoader.findResource(String)` as of today and will return an URL to a special protocol that is capable of reading a resource from the bundles classpath
> 2. If the URL is about to be opened will create a `WeavingURLConnection` with encapsulates a WovenResource initialized with a `Supplier<InputStream>` that can supply the data as it is done today
> 3. If there are WeavingHooks, pass that WovenResource to the Hooks and they can set the `Supplier<InputStream>` to a different one

So then it moves on to the next weaving hook and the `Supplier<InputStream>` may or may not be a previously called hooks supplier?  When do we call the actual weavers?  Right when the URLConnection is created or only when getInputStream is called?  I think it would have to be delayed to be as late as possible.  Otherwise you could get a URLConnection, weavers called and final Supplier<InputStream> obtained ... non-specific time passes ... finally get the InputStream from the supplier.  At this point the final weaving hook may have been unregistered, we cannot hold a lock to prevent registration for some non-specific time waiting for the supplier to get called.  What happens then?  Do we throw an IOException or do we silently fallback, if we fallback do we go back to the previous hook to get its Supplier again?  or do we fall all the way back to the original content?

> 4. if later the openStream is called it simply delegates to the supplier
> 
> > Weaving is primarily about modifying code to weave in some cross-cutting behavior.
> 
> And the same can apply to resources as well, how is this different? Lets say I'd like AspektJ or similar to add automatic logging to every class from the bussinesslayer, why should I not be allowed to replace a variable in all logging.properties files or alike?
> 
I do wonder why such a thing is not available to go hand in hand with `java.lang.instrument.ClassFileTransformer`.  I think most popular runtime weaving technologies today use this but it doesn't appear that the demand for resource transforming has elevated to the point of them defining such a thing for Java SE.

> > Once a class is loaded by a class loader the byte array is no longer interesting
> 
> The same applies to URLConnections beside that they have a much shorter life-cycle and a higher chance of freeing resources again

I don't see how you can say URLConnections are short lived.  That is all dependent on the user.  It is valid for the user to stash the connection for as long as they like.

> 
> > Resources are fundamentally different. They are not transformed by a class loader into some immutable form (e.g. Class object)
> 
> It clearly depends how they are used, but I disagree that they are fundamentally different (they are just not that limited), if you take for example a function that reads icons from the classpath, these would be read once, decoded transferred into some kind of in-memory representation of an pixel array and then be used unless they are disposed again. The same will apply to config files, I'm not aware of any framework that constantly parses a textual config file over and over again, they will be read once and stored in an internal representation (e.g. Properties object).

Unfortunately a specification has to be prepared for what is done if the unusual thing happens.  Just because your use case only reads a resource once doesn't mean there isn't some other case.  Users will always surprise you, like if someone has a 1 gigabyte resource in their bundle.

> 
> > would either require them to be mutated on each request [...] or require the framework to cache the resulting woven resource to return in the future
> 
> I don't see any need for caching nor problematic access as there is no much difference to today, a framework might choose to cache resources (to prevent disk I/O) or read it every time it is requested. If transforming adds a penalty clearly depends on the type of transformation and is out of the scope of the framework or this request.

This is not out of scope for the request.  We would need to decide if a resource is transformed once and the result is always the same from that point forward (maybe until the bundle is refreshed) or each and every time.  I think transforming it each time is dangerous, but I also think caching is not desired because it will add to the complexity and resources the framework consumes.

> 
> > What about resources in non-package directories (e.g. /META-INF folder).
> 
> As described above I would only consider resources that are loaded through classloaders.

I don't see why only class loader resources are considered.  My main point is mete-data resources read by the framework cannot be woven because it would happen far too late and we cannot let things like the bundle headers change dynamically.  But why would you not want to be able to weave DS component XML files?  I'm sure someone can think up an interesting case that should allow one to transform DS service component XML definitions on the fly.  I'm also not sure I would think it is a wise idea to add that to the specification though.  I also know there has been use cases in the past where the old equinox transformer project would allow things like `plugin.xml` resources to be transformed (which are bundle resources not class loader resources).

But Equinox transformer project has long been deprecated and dead because we didn't have any good usecases driving the interest and development of it.

> 
> > It could get confusing to differentiate woven resources obtained by the class loader and their corresponding resources that may be obtained by Bundle.getEntry without going through the class loader. I assume Bundle.getEntry resources would never be woven.
> 
> As of today, all these Bundle.getXXX methods are well described whether or not the BundleClasslaoder is used or not so there actually already is a difference. I would simply add a comment to those not using the classloader that ResourceWeaving would not happen.

I would not see the need to make that distinction.

> 
> > So I don't think this is a good idea.
> 
> Let me know your concerns so we can discuss them, from the once mentions until now many mechanism we find in modern OSGi Frameworks should have never been included (e.g. subsystems and alike), this one is really nothing much to harm...

That is my concerns.  We did have a specification call today to discuss this and others had some concerns also.  We just didn't write them down during the call when making the decision to close this.

**Comment by laeubi** [comment]
> But if weavers are involved then the content read may change over time as weavers are registered/unregistered.

Yes, but that's the same with many aspects in OSGi (All kind of Hooks including WeavingHooks, ConfigPlugin, ...) and the WeavingHook specification suggest to refresh the packages to make sure it weaves all classes as it might have missed (some) classes or woven classes remain afterwards.

That's why I wrote I won't require any caching or special handling from the framework to handle this, this is completely in control of the ResourceWeavingHook.

> When do we call the actual weavers? Right when the URLConnection is created or only when getInputStream is called?

I think it would make most sense to delay this until getInputStream is called (as code might just want to query if a resource exits instead of reading it), but that's a bit of an detail that could be decided in the specification process and backed with some POC code and how we like to handle the code where a resource is hidden (e.g.that null should be returned or IOException is thrown on open ...)

A ResourceHooks has to make sure that the supplier is either stateless or can handle bundle-shutdown (e.g. it could simply not do anything if required services are gone) but that's also nothing special in OSGi world as even services returned can go at any time and the code has to handle that case.

But one should also be realistic, of course theoretically "non-specific time passes" but that's not the usual use-case, the usual case (99%) a resource URL is fetched and immediately read (if not the caller even uses the getStream() method directly) and even without weaving theoretically one can get a resource URL through a classloader and later the bundle is uninstalled.

> I do wonder why such a thing is not available to go hand in hand with ...

I have no idea, but there are many things in OSGi that are not in Java SE so I won't take this as normative. I also assume that it is not the plan of the OSGi to throw away WeavingHooks just because ClassFileTransformer is probably an alternative... Also in a traditional Java SE Application you mostly have some control over classloaders, while in OSGi (client) code the framework takes over this control, so if OSGi itself does not offer such a thing there is no chance for a plugin to provide such functionality by itself.

And that's why I opened this request as there is literally no other way to archive this without the help of the framework itself.

> I don't see why only class loader resources are considered.

Because of the problems one might see otherwise as reading of "non-lassloader" resources as you already have described. Beside that it is well defined when a classloader is created for a bundle and when it becomes accessible (together with its resources). And if we can get this (more restrictive) use-case it would be a first step. And even though OSGi specific services access resources through other ways (e.g. to read them when no classloader is present), most others simply use the classlaoder to read (embedded) resources.

>  But why would you not want to be able to weave DS component XML files?

If its possible and do not complicate things to much why not, but that's not the primary intent here, the scope for this is clearly for resource loaded through classloaders and why ClassWeavingHooks are related I think. But of course plugin.xml and alike would also be interesting, anyways I won't mind to exclude especially the Manifest from any resource interception.

>  We did have a specification call today to discuss this and others had some concerns also. We just didn't write them down during the call when making the decision to close this.

But closing this without giving an alternative (I don't see any beside asking each framework that loads resources to provide a hook) here to allow such use-cases if the framework does not support them is a bit frustration without giving clear reasons (or ask for clarification/narrowing the scope) beside "we don't find it useful" as.of course if it is not been considered useful at least for me I won't have opened this.

**Comment by laeubi** [comment]
> But Equinox transformer project has long been deprecated and dead because we didn't have any good usecases driving the interest and development of it.

The problem with any provider specific solution is that it hinders wide-adoption and thus the proposal to have a standardized solution.

**Comment by laeubi** [comment]
I found that Equinox has `ClassLoaderHook` that supports resource interception with ClassLoaderHook some pre/postFindResources methods that look similar to what I would require, at least there is even a idea to use that to implement the mediator spec without a need to rewrite classes:
- https://github.com/eclipse-equinox/equinox/pull/295

If something like this would have been standardized, such thing do not need to rely on internal equinox API...

**Comment by tjwatson** [comment]
> If something like this would have been standardized, such thing do not need to rely on internal equinox API...

@HannesWell can confirm.  His draft PR for https://github.com/eclipse-equinox/equinox/pull/295 is to look into possibilities to implement #372.  That is about building the service loader mediator into the framework itself.  At that point I expect a framework implementation to use any of its own internals to implement that as easily and efficiently as possible.

**Comment by laeubi** [comment]
I'm not sure I understand, weather or not the service is directly implemented in the framework does not mean it is mandated to use internal API. I just wanted to share a case where

1.  it is useful to intercept resource loads
2. there is currently support for it but only in a non standard way
3. at least one major implementation of the OSGi framework found it useful enough and it seem to be working

So if I bind my component/service/extender/... to equinox it is already possible but makes it unusable for any other standard compliant framework (e.g. Felix) or I need to write a dedicated solution for each of them.

**Comment by tjwatson** [comment]
> I'm not sure I understand

My point is a scenario for an implementation of #372 does not justify adding new APIs to the Framework for resource weaving since that specification is about building something directly into the Framework itself.

**Comment by laeubi** [comment]
The Resolver is "building something directly into the Framework itself" but still there is a Resolver Specification because what is useful for the framework is often useful for others and sharing such thing prevents having to reinvent the wheel (what even is not possible here at all).

**Comment by tjwatson** [comment]
> The Resolver is "building something directly into the Framework itself"

I disagree that these are similar.  The Resolver specification started out as a separate specification altogether and not part of the core Framework specification.  It was not until after many months of work by Richard, Karl, BJ, myself and others to make sure the resolver specification was usable by the framework itself did we finally decide to pull the Resolver specification into the core framework.

Regardless, my main worry is that introducing weaving of resources will help about 5 developers in the world doing super advanced things.  Yet it will be very complex to get it right for these 5 developers.  I don't mind making these 5 developers do framework specific things to avoid adding more complexity to the specification.
